Don't use exceptions

In my opinion, using exceptions for saying that the user has no saved credentials is pretty bad design. It's common to use `null` for this case. This won't let my error reporting tool freak out.

Another idea is a `isSupported` method to prevent an exception raised because of a not existing Google API. This would also be very helpful with different platforms.

I mean these lines where you catch the Platform Exception:
https://github.com/rekire/autologin_plugin/blob/44292526d884c37418dc6ae2f7f26dbd197c0b39/example/lib/main.dart#L29-L33

A Platform Exception is thrown if the user has no saved credentials (If I'm not wrong):
https://github.com/rekire/autologin_plugin/blob/44292526d884c37418dc6ae2f7f26dbd197c0b39/android/src/main/kotlin/eu/rekisoft/flutter/autologin/LoginHelper.kt#L33-L37
https://github.com/rekire/autologin_plugin/blob/44292526d884c37418dc6ae2f7f26dbd197c0b39/android/src/main/kotlin/eu/rekisoft/flutter/autologin/LoginHelper.kt#L50-L54
That's where I expected the returned null value instead of an exception because it's a quiet normal to not have saved credentials.

A Platform Exception is also a pretty general exception. So if anything else goes wrong the Platform Exception gets catched and no one gets informed about the problem. A workaround for this would be parsing the exception message, but it's more complicated than it could be. I suggest null as return value or create a special exception inside dart so only this "error" can get catched.

Then I got this exception because of a not existing google account which I would avoid with a `isSupported` method:
```
PlatformException(16, Error code 16 while loading the login data (16: No eligible accounts can be found.), eu.rekisoft.flutter.autologin.LoginHelper$GoogleApiError: Error code 16 while loading the login data (16: No eligible accounts can be found.)
	at eu.rekisoft.flutter.autologin.LoginHelper$loadLoginData$1.onComplete(LoginHelper.kt:64)
	at com.google.android.gms.tasks.zzj.run(Unknown Source:4)
	at android.os.Handler.handleCallback(Handler.java:873)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loop(Looper.java:193)
	at android.app.ActivityThread.main(ActivityThread.java:6669)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
Caused by: com.google.android.gms.common.api.ApiException: 16: No eligible accounts can be found.
	at com.google.android.gms.common.internal.ApiExceptionUtil.fromStatus(com.google.android.gms:play-services-base@@17.1.0:4)
	at com.google.android.gms.common.internal.zai.zaf(com.google.android.gms:play-services-base@@17.1.0:2)
	at com.google.android.gms.common.internal.zak.onComplete(com.google.android.gms:play-services-base@@17.1.0:6)
	at com.google.android.gms.common.api.internal.BasePendingResult.zaa(com.google.android.gms:play-services-base@@17.1.0:176)
	at com.google.android.gms.common.api.internal.BasePendingResult.setResult(com.google.android.gms:play-services-base@@17.1.0:135)
	at com.google.android.gms.internal.auth-api.zzl.zzc(com.google.android.gms:play-services-auth@@18.0.0:2)
	at com.google.android.gms.internal.auth-api.zzu.dispatchTransaction(com.google.android.gms:play-services-auth@@18.0.0:6)
	at com.google.android.gms.internal.auth-api.zzc.onTransact(com.google.android.gms:play-services-auth@@18.0.0:13)
	at android.os.Binder.execTransact(Binder.java:731)
)
```
I would also add the availability check you already have to the `isSupported` method.
https://github.com/rekire/autologin_plugin/blob/44292526d884c37418dc6ae2f7f26dbd197c0b39/android/src/main/kotlin/eu/rekisoft/flutter/autologin/LoginHelper.kt#L23

--- a/android/src/main/kotlin/eu/rekisoft/flutter/autologin/AutologinPlugin.kt
+++ b/android/src/main/kotlin/eu/rekisoft/flutter/autologin/AutologinPlugin.kt
@@ -78,6 +78,12 @@ public class AutologinPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
                     LoginHelper.saveLoginData(binding, call["username"], call["password"], { result.success(true) }, ::handleError)
                 }
             }
+            "disableAutoLogIn" -> {
+                tasks.add { binding ->
+                    debug("disableAutoLogIn()")
+                    LoginHelper.disableAutoLogIn(binding, result::success, ::handleError)
+                }
+            }
             else -> result.notImplemented()
         }
         executeTasks()

--- a/android/src/main/kotlin/eu/rekisoft/flutter/autologin/AutologinPlugin.kt
+++ b/android/src/main/kotlin/eu/rekisoft/flutter/autologin/AutologinPlugin.kt
@@ -59,7 +59,11 @@ public class AutologinPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
             }
         }
         when (call.method) {
-            "getPlatformVersion" -> result.success("Android ${android.os.Build.VERSION.RELEASE}")
+            "isPlatformSupported" ->
+                tasks.add { binding ->
+                    debug("isPlatformSupported()")
+                    LoginHelper.isPlatformSupported(binding)
+                }
             "getLoginData" ->
                 tasks.add { binding ->
                     debug("loadLoginData()")

--- a/android/src/main/kotlin/eu/rekisoft/flutter/autologin/AutologinPlugin.kt
+++ b/android/src/main/kotlin/eu/rekisoft/flutter/autologin/AutologinPlugin.kt
@@ -20,7 +20,7 @@ public class AutologinPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
     /// when the Flutter Engine is detached from the Activity
     private lateinit var channel: MethodChannel
     private var binding: ActivityPluginBinding? = null
-    private var onBound: ((ActivityPluginBinding) -> Unit)? = null
+    private val tasks: MutableList<(ActivityPluginBinding) -> Unit> = mutableListOf()
 
     override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
         channel = MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "autologin_plugin")
@@ -60,32 +60,29 @@ public class AutologinPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
         }
         when (call.method) {
             "getPlatformVersion" -> result.success("Android ${android.os.Build.VERSION.RELEASE}")
-            "getLoginData" -> {
-                onBound = { binding ->
+            "getLoginData" ->
+                tasks.add { binding ->
                     debug("loadLoginData()")
                     LoginHelper.loadLoginData(binding, { username, password ->
                         result.success(listOf(username, password))
                     }, ::handleError)
-                    onBound = null
                 }
-                binding?.let {
-                    onBound?.invoke(it)
-                    onBound = null
-                }
-            }
             "saveLoginData" -> {
                 operator fun MethodCall.get(arg: String): String = requireNotNull(argument(arg)) { "$arg was null" }
-                onBound = { binding ->
+                tasks.add { binding ->
                     debug("saveLoginData()")
                     LoginHelper.saveLoginData(binding, call["username"], call["password"], { result.success(true) }, ::handleError)
-                    onBound = null
-                }
-                binding?.let {
-                    onBound?.invoke(it)
-                    onBound = null
                 }
             }
             else -> result.notImplemented()
+        }
+        executeTasks()
+    }
+
+    private fun executeTasks() {
+        binding?.let { binding ->
+            tasks.forEach { task -> task.invoke(binding) }
+            tasks.clear()
         }
     }
 
@@ -99,14 +96,14 @@ public class AutologinPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
 
     override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
         debug("onReattachedToActivityForConfigChanges()")
-        onBound?.invoke(binding)
         this.binding = binding
+        executeTasks()
     }
 
     override fun onAttachedToActivity(binding: ActivityPluginBinding) {
         debug("onAttachedToActivity()")
-        onBound?.invoke(binding)
         this.binding = binding
+        executeTasks()
     }
 
     override fun onDetachedFromActivityForConfigChanges() {

--- a/android/src/main/kotlin/eu/rekisoft/flutter/autologin/AutologinPlugin.kt
+++ b/android/src/main/kotlin/eu/rekisoft/flutter/autologin/AutologinPlugin.kt
@@ -26,8 +26,7 @@ public class AutologinPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
         channel = MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "autologin_plugin")
         channel.setMethodCallHandler(this)
 
-        //flutterPluginBinding.flutterEngine.add
-        println("onAttachedToEngine")
+        debug("onAttachedToEngine")
     }
 
     // This static function is optional and equivalent to onAttachedToEngine. It supports the old
@@ -43,7 +42,7 @@ public class AutologinPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
         @JvmStatic
         fun registerWith(registrar: Registrar) {
             val channel = MethodChannel(registrar.messenger(), "autologin_plugin")
-            println("registerWith " + registrar.activity().javaClass.simpleName)
+            debug("registerWith " + registrar.activity().javaClass.simpleName)
             channel.setMethodCallHandler(AutologinPlugin())
         }
     }
@@ -63,7 +62,7 @@ public class AutologinPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
             "getPlatformVersion" -> result.success("Android ${android.os.Build.VERSION.RELEASE}")
             "getLoginData" -> {
                 onBound = { binding ->
-                    println("loadLoginData()")
+                    debug("loadLoginData()")
                     LoginHelper.loadLoginData(binding, { username, password ->
                         result.success(listOf(username, password))
                     }, ::handleError)
@@ -77,7 +76,7 @@ public class AutologinPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
             "saveLoginData" -> {
                 operator fun MethodCall.get(arg: String): String = requireNotNull(argument(arg)) { "$arg was null" }
                 onBound = { binding ->
-                    println("saveLoginData()")
+                    debug("saveLoginData()")
                     LoginHelper.saveLoginData(binding, call["username"], call["password"], { result.success(true) }, ::handleError)
                     onBound = null
                 }
@@ -99,13 +98,13 @@ public class AutologinPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
     }
 
     override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
-        println("onReattachedToActivityForConfigChanges()")
+        debug("onReattachedToActivityForConfigChanges()")
         onBound?.invoke(binding)
         this.binding = binding
     }
 
     override fun onAttachedToActivity(binding: ActivityPluginBinding) {
-        println("onAttachedToActivity()")
+        debug("onAttachedToActivity()")
         onBound?.invoke(binding)
         this.binding = binding
     }

--- a/android/src/main/kotlin/eu/rekisoft/flutter/autologin/LoginHelper.kt
+++ b/android/src/main/kotlin/eu/rekisoft/flutter/autologin/LoginHelper.kt
@@ -19,6 +19,8 @@ object LoginHelper {
     private const val saveRequestCode = 48233
     internal const val debugCalls = true
 
+    fun isPlatformSupported(binding: ActivityPluginBinding) = GoogleApiAvailability().isGooglePlayServicesAvailable(binding.activity) == ConnectionResult.SUCCESS
+
     fun loadLoginData(binding: ActivityPluginBinding, callback: (username: String, password: String?) -> Unit, error: (Exception?) -> Unit) {
         val availability = GoogleApiAvailability().isGooglePlayServicesAvailable(binding.activity)
         if (availability == ConnectionResult.SUCCESS) {

--- a/android/src/main/kotlin/eu/rekisoft/flutter/autologin/LoginHelper.kt
+++ b/android/src/main/kotlin/eu/rekisoft/flutter/autologin/LoginHelper.kt
@@ -21,7 +21,7 @@ object LoginHelper {
 
     fun isPlatformSupported(binding: ActivityPluginBinding) = GoogleApiAvailability().isGooglePlayServicesAvailable(binding.activity) == ConnectionResult.SUCCESS
 
-    fun loadLoginData(binding: ActivityPluginBinding, callback: (username: String, password: String?) -> Unit, error: (Exception?) -> Unit) {
+    fun loadLoginData(binding: ActivityPluginBinding, callback: (username: String?, password: String?) -> Unit, error: (Exception?) -> Unit) {
         val availability = GoogleApiAvailability().isGooglePlayServicesAvailable(binding.activity)
         if (availability == ConnectionResult.SUCCESS) {
             val request = CredentialRequest.Builder()
@@ -32,11 +32,7 @@ object LoginHelper {
                 if (it.isSuccessful) {
                     val username = it.result?.credential?.id
                     val password = it.result?.credential?.password
-                    if (username != null) {
-                        callback(username, password)
-                    } else {
-                        error(GoogleApiError(-1))
-                    }
+                    callback(username, password)
                 } else {
                     when (it.exception) {
                         is ResolvableApiException -> {
@@ -49,11 +45,7 @@ object LoginHelper {
                                 val username = credential?.id
                                 val password = credential?.password
                                 debug("username: $username, password: (${if (password.isNullOrBlank()) "not set" else "removed"})")
-                                if (username != null) {
-                                    callback(username, password)
-                                } else {
-                                    error(GoogleApiError(-1))
-                                }
+                                callback(username, password)
                             }
                             try {
                                 (it.exception as ResolvableApiException).startResolutionForResult(binding.activity, loginRequestCode)
@@ -107,6 +99,14 @@ object LoginHelper {
         }
     }
 
+    fun disableAutoLogIn(binding: ActivityPluginBinding, success: (Boolean) -> Unit, error: (Exception?) -> Unit) {
+        val available = GoogleApiAvailability().isGooglePlayServicesAvailable(binding.activity) == ConnectionResult.SUCCESS
+        if (available) {
+            Credentials.getClient(binding.activity).disableAutoSignIn()
+        }
+        success(available)
+    }
+
     private fun ActivityPluginBinding.addActivityResultListener(requestCodeFilter: Int, listener: (resultCode: Int, data: Intent?) -> Unit) =
             addActivityResultListener(object : PluginRegistry.ActivityResultListener {
                 override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
@@ -124,4 +124,4 @@ object LoginHelper {
                     ?: "no details given"})", cause)
 }
 
-internal inline fun debug(msg: String) = if(LoginHelper.debugCalls) println(msg) else Unit
+internal inline fun debug(msg: String) = if (LoginHelper.debugCalls) println(msg) else Unit

--- a/android/src/main/kotlin/eu/rekisoft/flutter/autologin/LoginHelper.kt
+++ b/android/src/main/kotlin/eu/rekisoft/flutter/autologin/LoginHelper.kt
@@ -17,7 +17,7 @@ import io.flutter.plugin.common.PluginRegistry
 object LoginHelper {
     private const val loginRequestCode = 48232
     private const val saveRequestCode = 48233
-    private const val debugCalls = true
+    internal const val debugCalls = true
 
     fun loadLoginData(binding: ActivityPluginBinding, callback: (username: String, password: String?) -> Unit, error: (Exception?) -> Unit) {
         val availability = GoogleApiAvailability().isGooglePlayServicesAvailable(binding.activity)
@@ -42,11 +42,11 @@ object LoginHelper {
                             // credentials and needs to pick one. This requires showing UI to
                             // resolve the read request.
                             binding.addActivityResultListener(loginRequestCode) { _, data ->
-                                if (debugCalls) println("onActivityResult(_,_,$data)")
+                                debug("onActivityResult(_,_,$data)")
                                 val credential = data?.getParcelableExtra<Credential>(Credential.EXTRA_KEY)
                                 val username = credential?.id
                                 val password = credential?.password
-                                if (debugCalls) println("username: $username, password: (${if (password.isNullOrBlank()) "not set" else "removed"})")
+                                debug("username: $username, password: (${if (password.isNullOrBlank()) "not set" else "removed"})")
                                 if (username != null) {
                                     callback(username, password)
                                 } else {
@@ -121,3 +121,5 @@ object LoginHelper {
             RuntimeException("Error code $error while loading the login data (${cause?.message
                     ?: "no details given"})", cause)
 }
+
+internal inline fun debug(msg: String) = if(LoginHelper.debugCalls) println(msg) else Unit

--- a/example/android/app/src/main/kotlin/eu/rekisoft/flutter/autologin/autologin_plugin_example/MainActivity.kt
+++ b/example/android/app/src/main/kotlin/eu/rekisoft/flutter/autologin/autologin_plugin_example/MainActivity.kt
@@ -1,11 +1,5 @@
 package eu.rekisoft.flutter.autologin.autologin_plugin_example
 
-import android.content.Intent
 import io.flutter.embedding.android.FlutterActivity
 
-class MainActivity: FlutterActivity() {
-
-    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
-        super.onActivityResult(requestCode, resultCode, data)
-    }
-}
+class MainActivity: FlutterActivity()

--- a/example/android/build.gradle
+++ b/example/android/build.gradle
@@ -6,7 +6,7 @@ buildscript {
     }
 
     dependencies {
-        classpath 'com.android.tools.build:gradle:3.5.0'
+        classpath 'com.android.tools.build:gradle:3.6.0-rc02'
         classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
     }
 }

--- a/example/android/gradle/wrapper/gradle-wrapper.properties
+++ b/example/android/gradle/wrapper/gradle-wrapper.properties
@@ -1,6 +1,6 @@
-#Fri Jun 23 08:50:38 CEST 2017
+#Mon Jul 06 13:47:09 CEST 2020
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-5.6.2-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.6.4-all.zip

--- a/example/lib/main.dart
+++ b/example/lib/main.dart
@@ -55,7 +55,7 @@ class _MyAppState extends State<MyApp> {
         floatingActionButton: FloatingActionButton(
           child: Icon(Icons.save),
           onPressed: () {
-            AutologinPlugin.saveLoginData(Credential.fromArgs("test", "test"));
+            AutologinPlugin.saveLoginData(Credential("test", "test"));
           },
         ),
       ),

--- a/example/lib/main.dart
+++ b/example/lib/main.dart
@@ -2,7 +2,6 @@ import 'dart:async';
 
 import 'package:autologin_plugin/autologin_plugin.dart';
 import 'package:flutter/material.dart';
-import 'package:flutter/services.dart';
 
 void main() {
   runApp(MyApp());
@@ -24,12 +23,12 @@ class _MyAppState extends State<MyApp> {
 
   // Platform messages are asynchronous, so we initialize in an async method.
   Future<void> initPlatformState() async {
+    Credential credentials = await AutologinPlugin.getLoginData();
     String username;
-    // Platform messages may fail, so we use a try/catch PlatformException.
-    try {
-      username = (await AutologinPlugin.getLoginData()).username;
-    } on PlatformException {
-      username = '<No account chosen>';
+    if (credentials != null) {
+      username = credentials.username;
+    } else {
+      username = "<nothing selected>";
     }
 
     // If the widget was removed from the tree while the asynchronous platform

--- a/lib/autologin_plugin.dart
+++ b/lib/autologin_plugin.dart
@@ -18,6 +18,10 @@ class AutologinPlugin {
   static Future<bool> saveLoginData(Credential credential) async {
     return await _channel.invokeMethod('saveLoginData', <String, dynamic>{'username': credential.username, 'password': credential.password});
   }
+
+  static Future<bool> disableAutoLogIn() async {
+    return await _channel.invokeMethod('disableAutoLogIn()');
+  }
 }
 
 class Credential {

--- a/lib/autologin_plugin.dart
+++ b/lib/autologin_plugin.dart
@@ -6,9 +6,8 @@ class AutologinPlugin {
   static const MethodChannel _channel =
       const MethodChannel('autologin_plugin');
 
-  static Future<String> get platformVersion async {
-    final String version = await _channel.invokeMethod('getPlatformVersion');
-    return version;
+  static Future<String> get isPlatformSupported async {
+    return await _channel.invokeMethod('isPlatformSupported');
   }
 
   static Future<Credential> getLoginData() async {

--- a/lib/autologin_plugin.dart
+++ b/lib/autologin_plugin.dart
@@ -13,7 +13,7 @@ class AutologinPlugin {
 
   static Future<Credential> getLoginData() async {
     final List<dynamic> data = await _channel.invokeMethod('getLoginData');
-    return Credential.fromArgs(data[0] as String, data[1] as String);
+    return Credential(data[0] as String, data[1] as String);
   }
 
   static Future<bool> saveLoginData(Credential credential) async {
@@ -25,10 +25,5 @@ class Credential {
   String username;
   String password;
 
-  Credential({this.username, this.password});
-
-  Credential.fromArgs(String username, String password) {
-    this.username = username;
-    this.password = password;
-  }
+  Credential(this.username, this.password);
 }

--- a/lib/autologin_plugin.dart
+++ b/lib/autologin_plugin.dart
@@ -3,15 +3,15 @@ import 'dart:async';
 import 'package:flutter/services.dart';
 
 class AutologinPlugin {
-  static const MethodChannel _channel =
-      const MethodChannel('autologin_plugin');
+  static const MethodChannel _channel = const MethodChannel('autologin_plugin');
 
-  static Future<String> get isPlatformSupported async {
+  static Future<bool> get isPlatformSupported async {
     return await _channel.invokeMethod('isPlatformSupported');
   }
 
   static Future<Credential> getLoginData() async {
     final List<dynamic> data = await _channel.invokeMethod('getLoginData');
+    if(data[0] == null) return null;
     return Credential(data[0] as String, data[1] as String);
   }
 

